lib/gc/Transforms/GPU/AllocsToSLM.cpp

```diff
@@ -0,0 +1,97 @@
+//===- AllocsToSLM.cpp - A pass adding shared mem-space attr ----*- C++ -*-===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "gc/Transforms/Passes.h"
+
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/GPU/TransformOps/Utils.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/IR/Dialect.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#include <numeric>
+#include <optional>
+
+using namespace mlir;
+using namespace mlir::gc;
+
+namespace mlir {
+namespace gc {
+#define GEN_PASS_DEF_ALLOCSTOSLM
+#include "gc/Transforms/Passes.h.inc"
+} // namespace gc
+} // namespace mlir
+
+namespace {
+
+bool isInGpuLaunch(mlir::Operation *op) {
+  auto launchOp = op->getParentOfType<gpu::LaunchOp>();
+  return launchOp != nullptr;
+}
+
+bool hasAssignedMemSpace(mlir::Value value) {
+  if (auto memrefType = value.getType().dyn_cast<mlir::MemRefType>()) {
```

Comment:
```suggestion
  if (auto memrefType = dyn_cast<MemRefType>(value.getType())) {
```

Type.dyn_cast() is deprecated:
```C++
  [[deprecated("Use mlir::dyn_cast<U>() instead")]]
  U dyn_cast() const;
```